lib/cylc/exceptions.py

```diff
@@ -15,17 +15,168 @@
+    If this isn't possible raise UserInputError.
+
     """
-    Attributes:
-        message - what the problem is.
+
+
+class LogAnalyserError(CylcError):
+    """Exception for issues scraping Cylc suite log files."""
+
+
+class CylcConfigError(CylcError):
+    """Generic exception to handle an error in a Cylc configuration file.
+
+    TODO:
+        * reference the configuration element causing the problem
+
     """
+
+
+class SuiteConfigError(CylcConfigError):
+    """Exception for configuration errors in a Cylc suite configuration."""
+
+
+class GlobalConfigError(CylcConfigError):
+    """Exception for configuration errors in a Cylc global configuration."""
+
+
+class GraphParseError(SuiteConfigError):
+    """Exception for errors in Cylc suite graphing."""
+
+
+class TriggerExpressionError(GraphParseError):
+    """Trigger expression syntax issue."""
+
+
+class TaskProxySequenceBoundsError(CylcError):
+    """Error on TaskProxy.__init__ with out of sequence bounds start point."""
+
     def __init__(self, msg):
-        Exception.__init__(self, msg)
-        self.msg = msg
+        CylcError.__init__(
+            self, 'Not loading %s (out of sequence bounds)' % msg)
+
+
+class ParamExpandError(SuiteConfigError):
+    """Exception for errors in Cylc parameter expansion."""
+
+
+class SuiteEventError(CylcError):
+    """Exception for errors in Cylc event handlers."""
+
+
+class SuiteServiceFileError(CylcError):
+    """Exception for errors related to suite service files."""
+
+
+class TaskRemoteMgmtError(CylcError):
+    """Exceptions initialising suite run directory of remote job host."""
+
+    MSG_INIT = '%s: initialisation did not complete:\n'  # %s owner_at_host
+    MSG_SELECT = '%s: host selection failed:\n'  # %s host
+    MSG_TIDY = '%s: clean up did not complete:\n'  # %s owner_at_host
+
+    def __str__(self):
+        msg, (host, owner), cmd_str, ret_code, out, err = self.args
+        if owner:
+            owner_at_host = owner + '@' + host
+        else:
+            owner_at_host = host
+        ret = (msg + 'COMMAND FAILED (%d): %s\n') % (
+            owner_at_host, ret_code, cmd_str)
+        for label, item in ('STDOUT', out), ('STDERR', err):
+            if item:
+                for line in item.splitlines(True):  # keep newline chars
+                    ret += 'COMMAND %s: %s' % (label, line)
+        return ret
+
+
+class TaskDefError(SuiteConfigError):
+    """Exception raise for errors in TaskDef initialization."""
+
+
+class ClientError(CylcError):
 
     def __str__(self):
-        return repr(self.msg)
+        ret = 'Request returned error: %s' % self.args[0]
+        if len(self.args) > 1 and self.args[1]:
+            # append server-side traceback if appended
+            ret += '\n' + self.args[1]
+        return ret
+
+
+class ClientTimeout(CylcError):
+    pass
+
+
+class CyclingError(CylcError):
+    pass
+
+
+class CyclerTypeError(CyclingError):
```

Comment:
Do we need super-specialised exceptions like this?

Reply:
Presumably a good reason for a "super-specialised exception" is if the calling code needs to catch and respond specifically to that specific exception; otherwise it's pointless.